R/LSproblemofpointscommon.R

```diff
@@ -1,227 +1,234 @@
-sum_p_n <- function(y, a){
-  dDirichletnmultinom <- function(y, a){
+## This is a function calculating the probability mass function 
+# of Dirichlet negative multinomial distribution
+
+## Input:
+# y: a vector containing the number of trials each player needs to win
+# a: a vector containing all the parameters for the Dirichlet multinomial distribution, indicating the competence of each player
+
+pDirichletNegMultinom <- function(y, a){
+  dDirichletNegMultinom <- function(y, a){
```

Comment:
Two things here.

1. Please don't define functions inside other functions, unless they are lambdas (e.g., `sapply(list, function(x) {...})`).
2. What are typical values for `y` and `a`? Because the code as written here will very quickly overflow to `Inf`. For example,
```r
y <- a <- 20:25
prod(gamma(y + a))
# Inf
sum(lgamma(y + a))
# 752.6915
```
when working with probability distributions it is common to do all work on the log scale to avoid overflowing or underflowing. In that case that would become:
```r
dDirichletNegMultinom2 <- function(y, a, log = FALSE) {
  
  ySum <- sum(y)
  aSum <- sum(a)
  
  lp1 <- lgamma(ySum) - (lgamma(y[1]) + sum(lgamma(y+1)) - lgamma(y[1]+1))
  lp2 <- lgamma(aSum) - sum(lgamma(a))
  lp3 <- sum(lgamma(y + a)) - lgamma(ySum + aSum)

  logDensity <- lp1 + lp2 + lp3
  if (log)
    return(logDensity)
  else
    return(exp(logDensity))
    
}
```
The original `dDirichletNegMultinom` underflows when called like `dDirichletNegMultinom(1:20, 1:20)` but `dDirichletNegMultinom2` does not. Likewise, you could compute `pCalculated` with the logSumExp trick, but it kinda depends on typical values of `y` and `a` whether this is useful.

Reply:
Great point! Though typicall `y` and `a` are vectors containing small numbers with length 2 or 3, it is better to use the log scale to avoid overflowing to `Inf`.

Reply:
I just moved `dDirichletNegMultinom` out of `pDirichletNegMultinom`